src/audio/codec_adapter/codec/waves.c

```diff
@@ -0,0 +1,756 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2020 Waves Audio Ltd. All rights reserved.
+//
+// Author: Oleksandr Strelchenko <oleksandr.strelchenko@waves.com>
+//
+#include "sof/audio/codec_adapter/codec/generic.h"
+#include "sof/audio/codec_adapter/codec/waves.h"
+
+#include "MaxxEffect/MaxxEffect.h"
+#include "MaxxEffect/MaxxStream.h"
+#include "MaxxEffect/MaxxStatus.h"
+#include "MaxxEffect/Initialize/MaxxEffect_Initialize.h"
+#include "MaxxEffect/Process/MaxxEffect_Process.h"
+#include "MaxxEffect/Process/MaxxEffect_Reset.h"
+#include "MaxxEffect/Control/RPC/MaxxEffect_RPC_Server.h"
+#include "MaxxEffect/Control/Direct/MaxxEffect_Revision.h"
+
+#define MAX_CONFIG_SIZE_BYTES (8192)
+#define NUM_IO_STREAMS (1)
+
+struct waves_codec_data {
+	uint32_t                sample_rate;
+	uint32_t                buffer_bytes;
+	uint32_t                buffer_samples;
+	uint32_t                sample_size_in_bytes;
+	uint64_t                reserved;
+
+	MaxxEffect_t            *effect;
+	uint32_t                effect_size;
+	MaxxStreamFormat_t      i_format;
+	MaxxStreamFormat_t      o_format;
+	MaxxStream_t            i_stream;
+	MaxxStream_t            o_stream;
+	MaxxBuffer_t            i_buffer;
+	MaxxBuffer_t            o_buffer;
+	uint32_t                response_max_bytes;
+	uint32_t                request_max_bytes;
+	void                    *response;
+};
+
+enum waves_codec_params {
+	PARAM_NOP = 0,
+	PARAM_MESSAGE = 1,
+	PARAM_REVISION = 2
+};
+
+static int32_t sample_format_convert_to_bytes(MaxxBuffer_Format_t format)
+{
+	/* converts MaxxBuffer_Format_t to number of bytes it requires */
+	int32_t res;
+
+	switch (format) {
+	case MAXX_BUFFER_FORMAT_Q1_15:
+		res = sizeof(uint16_t);
+		break;
+	case MAXX_BUFFER_FORMAT_Q1_23:
+		res = 3; /* 3 bytes */
+		break;
+	case MAXX_BUFFER_FORMAT_Q9_23:
+	case MAXX_BUFFER_FORMAT_Q1_31:
+	case MAXX_BUFFER_FORMAT_Q5_27:
+		res = sizeof(uint32_t);
+		break;
+	case MAXX_BUFFER_FORMAT_FLOAT:
+		res = sizeof(float);
+		break;
+	default:
+		res = -EINVAL;
+		break;
+	}
+	return res;
+}
+
+static MaxxBuffer_Format_t format_convert_sof_to_me(enum sof_ipc_frame format)
+{
+	/* converts enum sof_ipc_frame to MaxxBuffer_Format_t */
+	MaxxBuffer_Format_t res;
```

Comment:
this should be placed above the function definition:

e.g 
```
/* converts enum sof_ipc_frame to MaxxBuffer_Format_t */
static MaxxBuffer_Format_t format_convert_sof_to_me(enum sof_ipc_frame format)
{
	MaxxBuffer_Format_t res;
```

Reply:
moved all such comments above function definition

Reply:
@stolx thanks, this change looks good. I'm marking this conversation as resolved.